torch/monitor: add C++ events and handlers

d4l3k opened the pull request:
Differential Revision: D32606547

kiukchung commented on torch/csrc/monitor/counters.h:
```diff
@@ -104,88 +103,138 @@ class Stat {
+      e.metadata[key.str()] = kv.second;
     }
+
+    logEvent(e);
```
having counters get published like Events is a fair choice for structured logging services (e.g. Scuba), but for simple timeseries based metric logging services (e.g. ods), the OdsEventHandler would have to just drop a bunch of fields (e.g. type) and just read off `event.metadata["foo.bar.AVG"] -> value`. But how does it know which key in the metadata map to read? would it iterate through all of them and only publish the ones where the value is a "number" type?

kiukchung commented on torch/csrc/monitor/events.cpp:
```diff
@@ -0,0 +1,61 @@
+namespace monitor {
+
+namespace {
+class EventHandlers {
```
Since we don't have the "ticker" we may miss to publish the last batch of events/metrics. Maybe add in a destructor that calls `logEvent()` for the final time or add an explicit `close()` method that the user has to call at the end of their program for now.

d4l3k replied:
done

kiukchung reviewed (commented):
LGTM overall, left some "comment-questions"

d4l3k commented on torch/csrc/monitor/events.h:
```diff
@@ -0,0 +1,38 @@
+
+using metadata_value_t = c10::variant<std::string, double, int64_t, bool>;
+
+struct Event {
```
Per the RFC it has a type which can be used to encode both. You could make it be `torchx.RunnerEvent`

kiukchung replied:
I see, its fine if we encode it as a single variable (type) and use a well-known delimiter (".").  The issue is that if the log archiving service offers more columns (e.g type, group, name), then we are mapping type -> multiple columns. 

In this case, we'd be splitting `torch.monitor.Event.type` in the EventHandler, so there is an implicit assumption about what the type string will split into depending on the Event Handler. 

For example if type = "torch.monitor.Stat.foobar.AVG" and the event sink offers  "type", "group", and "name" columns. It could decide to log `(torch, monitor.Stat, foobar.AVG)`, or `(torch.monitor, Stat, foobar.AVG)` or (torch.monitor, stat.foobar, AVG)`, etc...

I think this is fine, but does require some "best-practice" guidelines on how one should implement an EventHandler in this case.

d4l3k replied:
added comments on how to best use these events and handler interfaces

d4l3k committed:
torch/monitor: add C++ events and handlers (#68783)

Summary:
This adds a C++ event handler corresponding to the Python one mentioned in the RFC.

This changes the counters a bit to all be push driven instead of being polled. The two window types are "fixed count" and "interval". One is based off the number of logged events and the other is based off of time windows. There's currently no active ticker for interval so it needs a regular stream of events to ensure events are produced. A follow up diff can add support for things like HHWheel / simple ticker.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/68783

Test Plan: buck test //caffe2/test/cpp/monitor:monitor

Reviewed By: kiukchung

Differential Revision: D32606547

fbshipit-source-id: 5b3b4544d07766a2a376487fef1bc0ab08c84f0f

suo commented:
This diff broke Mac builds on master: https://github.com/pytorch/pytorch/runs/4495465412?check_suite_focus=true, I am reverting.

You can use ciflow to invoke mac builds on your PR.
